### ISSUE TITLE
Overlap combinations

### ISSUE THREAD

**Issue by mbhall88** [lh3/minimap2#1246]
This is a hybrid question/comment. It is something that I couldn't find documented anywhere, but please point me to it if it is. I am selfishly using this as a self-documentation issue for my future self (with the hope one other person in the future will find it useful).

I have been working with all-v-all overlaps (from a FASTQ).

My task was to count the number of overlaps each read has with the rest of the reads.

I (naively) thought that if I iterate over each row in the PAF file, keeping a counter for each query read (qname; column 1), this would give me the same number of overlaps for each read as if I used the target name (tname; column 6).

To my surprise, doing it both ways gave me very different overlap counts for many reads. This is a plot that shows the difference in the number of overlaps for a read depending on whether you use the qname or tname for each row.

![image](https://github.com/user-attachments/assets/98da7006-f2a0-4f02-ba4e-e40cc17f4c7c)

Essentially, this says that for most reads the difference is not zero.

## tl;dr

the query and target name pairs are all pairwise combinations without replacement. That is, if r1 and r2 overlap, there will only be one entry in the PAF file with, for example, r1 as qname and r2 as tname; you won't find r1 as tname and r2 as qname. However, there *can* be self-v-self overlaps - i.e., r1 in qname and r1 in tname.

So to get the correct number of overlaps for each entry in the PAF, increment the overlap count for the qname *and* tname (provided they aren't the same read and you haven't seen that pair already) as you won't see the inverse, and therefore relying on the qname or tname alone will underestimate the number of overlaps.

Please correct me if I am off here Heng.

**Comment by lh3** [comment]
Are you looking for `--dual=yes`? There is also `-D` but it is not intended for read overlapping. See manpage.

**Comment by mbhall88** [comment]
Ah maybe. I guess it wasn't super clear to me exactly what that meant for my situation. So seems I've essentially described the functionality of `--dual=no`? This is the default when using `-x ava-ont` looking at the manpage. Also, seems `-D` is also used by default as it is implied by `-X` which is implied by `-x ava-ont`, why is it used for the ava presets when you say it is not intended for read overlapping?

**Comment by lh3** [comment]
I mainly point you to the two options. Look up the manpage and decide how to use.
